tooling/minitest/src/tests/compute_align.rs

```diff
@@ -0,0 +1,198 @@
+
+        f.exit();
+        p.finish_function(f)
+    };
+
+    let p = p.finish_program(f);
+    assert_stop::<BasicMem>(p);
+}
+
+/// Tests compute_align for pointers.
+#[test]
+fn align_of_ptr() {
+    let mut p = ProgramBuilder::new();
+
+    let f = {
+        let mut f = p.declare_function();
+
+        assume_align_of_ty_conv::<&u8>(&mut f, 8);
+        assume_align_of_ty_conv::<&bool>(&mut f, 8);
+        assume_align_of_ty_conv::<&()>(&mut f, 8);
+        assume_align_of_ty_conv::<&[u8]>(&mut f, 8);
+        assume_align_of_ty_conv::<&mut [u16]>(&mut f, 8);
+        assume_align_of_ty_conv::<&[u8; 2]>(&mut f, 8);
+
+        f.exit();
+        p.finish_function(f)
+    };
+
+    let p = p.finish_program(f);
+    assert_stop::<BasicMem>(p);
+}
+
+/// Tests compute_align for zero sized types.
+#[test]
+fn align_of_zst() {
+    let mut p = ProgramBuilder::new();
+
+    let f = {
+        let mut f = p.declare_function();
+
+        assume_align_of_ty(&mut f, 1, <()>::get_type());
+        assume_align_of_ty(&mut f, 4, <[u32; 0]>::get_type());
+
+        f.exit();
+        p.finish_function(f)
+    };
+
+    let p = p.finish_program(f);
+    assert_stop::<BasicMem>(p);
+}
+
+/// Tests compute_align for tuple types.
+#[test]
+fn align_of_struct() {
+    let mut p = ProgramBuilder::new();
+
+    let f = {
+        let mut f = p.declare_function();
+        assume_align_of_ty(
+            &mut f,
+            8,
+            tuple_ty(
+                &[(size(0), <u64>::get_type()), (size(8), <u32>::get_type())],
+                size(16),
+                align(8),
+            ),
+        );
+        f.exit();
+        p.finish_function(f)
+    };
+
+    let p = p.finish_program(f);
+    assert_stop::<BasicMem>(p);
+}
+
+/// Tests compute_align for slices.
+#[test]
+fn align_of_slice() {
+    let mut p = ProgramBuilder::new();
+
+    let f = {
+        let mut f = p.declare_function();
+
+        // Make arrays, get slice pointers to them and get their size
+        let arr = f.declare_local::<[u32; 3]>();
+        f.storage_live(arr);
+        let slice_ptr = construct_wide_pointer(
+            addr_of(arr, <&()>::get_type()),
+            const_int(3_usize),
+            <&[u32]>::get_type(),
+        );
+        f.assume(eq(
+            compute_align(<[u32]>::get_type(), get_metadata(slice_ptr)),
+            const_int(4_usize),
+        ));
+
+        let arr = f.declare_local::<[u32; 0]>();
+        f.storage_live(arr);
+        let slice_ptr = construct_wide_pointer(
+            addr_of(arr, <&()>::get_type()),
+            const_int(0_usize),
+            <&[u32]>::get_type(),
+        );
+        f.assume(eq(
+            compute_align(<[u32]>::get_type(), get_metadata(slice_ptr)),
+            const_int(4_usize),
+        ));
+
+        let arr = f.declare_local::<[u8; 312]>();
+        f.storage_live(arr);
+        let slice_ptr = construct_wide_pointer(
+            addr_of(arr, <&()>::get_type()),
```

Comment:
Why `&()`? Isn't this a `&[u8; 312]`?

Reply:
Fair, it would be more accurate, but we're casting again immediately afterwards, therefore this just makes it more verbose. Also this notion is already in use in slice tests.

Reply:
In just find the wrong type makes the test harder to read. So if it's not too hard, would be good to use more accurate types.